### PR TITLE
fix #947 and #1004 Add getContext() to Signal from materialize/doOnEach

### LINKED ISSUE #947: Support for simpler use of Context on different signals
Ideally should be able to do simple operations using context on next/error signals such as set/reset current thread context classloader or maybe set a ThreadLocal for some legacy code.

Currently, this can be done fairly easily for `onNext` signals, but in cases where there is no need to do anything with the event, it preferred if it was possible to use the current context in `doOnNext` or `doOnEach`.

For error signals, on the other hand, there is currently no way, that I could find, to do something simple using the current context on error signals.

It may not make sense to have a context-aware version of all onX methods, but by having a `doOnEach` version that receives the current context at least these use case could be achieved easily.

Currently, need to do something like this:

```
.transform(doOnNextOrErrorWithContext(context -> context.getOrEmpty(TCCL_REACTOR_CTX_KEY).ifPresent(cl -> currentThread().setContextClassLoader((ClassLoader) cl))))
.transform(next)
.transform(doOnNextOrErrorWithContext(context -> context.getOrEmpty(TCCL_ORIGINAL_REACTOR_CTX_KEY).ifPresent(cl -> currentThread().setContextClassLoader((ClassLoader) cl)))));


  private Function<? super Publisher<CoreEvent>, ? extends Publisher<CoreEvent>> doOnNextOrErrorWithContext(Consumer<Context> contextConsumer) {
    return lift((scannable, subscriber) -> new CoreSubscriber<CoreEvent>() {

      private Context context = subscriber.currentContext();

      @Override
      public void onNext(CoreEvent event) {
        contextConsumer.accept(context);
        subscriber.onNext(event);
      }
      @Override
      public void onError(Throwable throwable) {
        contextConsumer.accept(context);
        subscriber.onError(throwable);
      }

      @Override
      public void onComplete() {
        subscriber.onComplete();
      }

      @Override
      public Context currentContext() {
        return context;
      }

      @Override
      public void onSubscribe(Subscription s) {
        subscriber.onSubscribe(s);
      }
    });
  }
```

But should really be able to do something like this:
```
.doOnEachWithContext((s, context) -> {
    context.getOrEmpty(TCCL_REACTOR_CTX_KEY).ifPresent(cl -> currentThread().setContextClassLoader((ClassLoader) cl))))
}
.transform(next)
.doOnEachWithContext((s, context) -> {
    context.getOrEmpty(TCCL_REACTOR_CTX_KEY).ifPresent(cl -> currentThread().setContextClassLoader((ClassLoader) cl))))
}
```

### PR DESCRIPTION
**Pull request by simonbasle**
- filled for doOnEach
 - not filled for materialize for now

### CODE REVIEW

**Comment by smaldini** [comment]
Use currentContext vs getContext -
Eventually cache currentContext in FluxOnEach.

**Comment by dfeist** [comment]
@simonbasle @smaldini In my custom implementation for this as shown in the github issue, not caching the context produced a not insignificant overhead. Unless it's particularly hard to cache this, I'd try to include that in this PR.

**Comment by simonbasle** [comment]
@smaldini regarding the naming, I kept it consistent with other methods in `Signal` (notably `getType()`). The subscriber would both implement `getContext()` and `currentContext()`, of course...
Added commit to align `materialize` with this change and cache the `Context` in both cases, as @dfeist suggested there was a significant benefit in doing so.

File: reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java [hunk 1]
```diff
@@ -33,14 +34,16 @@
 
 	private static final long serialVersionUID = -2004454746525418508L;
 
+	private final Context    context;
```
Comment (smaldini) [inline reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java]: maybe transient, context is not serializable right now (should it?)

**Pull request merged**